packages/react-graphql/src/hooks/graphql-document.ts

```diff
@@ -0,0 +1,55 @@
+import {useState, useEffect, useCallback} from 'react';
+import {OperationVariables} from 'apollo-client';
+import {DocumentNode} from 'graphql-typed';
+
+import {AsyncQueryComponentType} from '..';
+
+export default function useGraphQLDocument<
+  Data = any,
+  Variables = OperationVariables,
+  DeepPartial = {}
+>(
+  documentOrComponent:
+    | DocumentNode<Data, Variables>
+    | AsyncQueryComponentType<Data, Variables, DeepPartial>,
+) {
+  const [document, setDocument] = useState<DocumentNode<
+    Data,
+    Variables
+  > | null>(() => {
+    if (isDocumentNode(documentOrComponent)) {
+      return documentOrComponent;
+    } else {
+      return documentOrComponent.resolved;
+    }
+  });
+
+  const loadDocument = useCallback(
+    async () => {
+      if (!isDocumentNode(documentOrComponent)) {
+        try {
+          const resolved = await documentOrComponent.resolve();
```

Comment:
arg...I would have assume that react have this behaviour imbedded in all the hooks now. (don't know why I thought that...)

Reply:
@lemonmade curious as to why you are using a `ref` in the hook above instead of a variable? I can see how `useRef` can be useful if the value being assigned was an object since we can re-use the same reference without creating new object references, however, if it is a boolean,  it should be fine to use `let mounted` ?

Reply:
> I think we need to check whether the component is mounted before doing this?

I likely don't have context, but why do we need to check for `mounted` ? Won't this function `useGraphQLDocument` not called at all if the component using it is not mounted? Feels like I am missing context 🤔

Reply:
Yes, the other option here is just to `let mounted = true;`, then set it to `false` in the callback, then check that variable in scope after the async operation. The hook just wraps that logic up (the only way to wrap it up in a hook is to do it with a `ref`, you can't really store a persisted value any other way).

Yes, `useEffect` only gets called when the component is mounted, but that doesn't mean it will end while the component is still mounted. If you kick off an async operation (as is done here), there is no requirement for React to wait for it to be done while the component is still mounting (it would be catastrophic if the UI didn't update until the promise resolved, and it would make no sense to keep the component mounted beyond when a parent unmounts it).

Reply:
so what to know is this `setState` is after an `await`
the component will be mount when the async method start, but may not be mounted by the time it finished.

`Ref` is normally uses for accessing the dom, but in hooks land, it is also the react recommended way to [keeping value](https://reactjs.org/docs/hooks-faq.html#is-there-something-like-instance-variables) you will normally keep around using property variable in class component.

Since hook turn the function itself into a component, just using a variables that can be access through out the different hook can have unexpected result.

Reply:
The is literally the exact same problem solved by having a `mounted` instance variable on a class component.

Reply:
Makes sense. Thanks to you both!

As for the `mounted` ref, an ideal solution would have been to have some kind of `cancel/abort` for `loadDocument` which we return from the `useEffect` so that it cancels the document load if the component unmounts. Having said that, I can't think of a reliable way to do so for this use case.

Reply:
Yes, it would definitely be preferable, we can potentially wrap in a cancellable promise as an alternative (I suggested a mounted ref because I had already written it, a cancellable promise is a little trickier)